File: python/unique_paths.py
class Solution(object):
    # O(n * m) time | O(n * m) space
    def uniquePaths(self, m, n):
        """
        :type m: int
        :type n: int
        :rtype: int
        """
        dp = [[0] * (n + 1) for _ in range(m + 1)]
        dp[m - 1][n - 1] = 1

        for i in range(m - 1, -1, -1):
            for j in range(n - 1, -1, -1):
                dp[i][j] += dp[i + 1][j] + dp[i][j + 1]
        return dp[0][0]

solution = Solution()
print(solution.uniquePaths(3, 7))

File: src/main/java/june_2025/UniquePaths.java
package june_2025;

public class UniquePaths {

    public static void main(String[] args) {
        int m = 3;
        int n = 7;

        UniquePaths uniquePaths = new UniquePaths();
        int result = uniquePaths.uniquePaths(m, n);
        System.out.println(result);
    }

    // O(n*m) time | O(n*m) space
    public int uniquePaths(int m, int n) {
        int[][] dp = new int[m][n];

        for (int r = 0; r < dp.length; r++) {
            dp[r][0] = 1;
        }
        for (int c = 0; c < dp[0].length; c++) {
            dp[0][c] = 1;
        }

        for (int r = 1; r < dp.length; r++) {
            for (int c = 1; c < dp[r].length; c++) {
                dp[r][c] = dp[r - 1][c] + dp[r][c - 1];
            }
        }

        return dp[dp.length - 1][dp[0].length - 1];
    }

}
